File: frontend/src/types/task.ts
export interface Task {
  id: string;
  project_id: string;
  user_id?: string;
  project_name?: string;
  title: string;
  description?: string;
  status: TaskStatus;
  priority: TaskPriority;
  assigned_agent_id?: string;
  assigned_agent?: string;
  agent_name?: string;
  workflow_id?: string;
  workflow_name?: string;
  workflow_snapshot?: Record<string, unknown>;
  schedule_type?: 'once' | 'cron' | 'interval';
  schedule_config?: Record<string, unknown>;
  config_overrides?: TaskConfigOverride[];
  created_by: string;
  input?: Record<string, unknown>;
  output?: Record<string, unknown>;
  human_context?: string;
  human_code_snippet?: string;
  human_attachments?: string[];
  started_at?: string;
  completed_at?: string;
  created_at: string;
  updated_at: string;
  /** Gateway 费用（美元） */
  cost_usd?: number;
  /** Gateway 重试次数 */
  retry_count?: number;
  /** Gateway 最大重试次数 */
  max_retries?: number;
}

export type TaskStatus =
  | 'pending'
  | 'running'
  | 'paused'
  | 'scheduled'
  | 'completed'
  | 'failed'
  | 'cancelled'
  | 'pending_human';

export type TaskPriority = 'low' | 'medium' | 'high' | 'critical';

export interface TaskConfigOverride {
  workflow_node_id: string;
  agent_type_id?: string;
  config_override: Record<string, unknown>;
}

export interface ScheduleConfig {
  type: 'immediate' | 'cron' | 'interval';
  cron_expression?: string;
  interval_seconds?: number;
}

export interface TaskAgentConfig {
  id: string;
  task_id: string;
  workflow_node_id: string;
  agent_type_id?: string;
  config_override: Record<string, unknown> | null;
  created_at: string;
  updated_at: string;
}

/** 三层级树节点 — 项目级 */
export interface TaskTreeProject {
  project_id: string;
  project_name: string;
  running_count: number;
  completed_count: number;
  failed_count: number;
  tasks: TaskTreeTask[];
}

/** 三层级树节点 — 任务级 */
export interface TaskTreeTask {
  id: string;
  project_id: string;
  title: string;
  status: Task['status'];
  priority: Task['priority'];
  assigned_agent_id?: string;
  agent_name?: string;
  progress: number;
  started_at?: string;
  updated_at: string;
  created_at: string;
  logs?: Array<{ level: string; message: string; timestamp: string }>;
  output_files?: Array<{ name: string; path: string }>;
}

/** Workflow node from backend definition */
export interface WorkflowNode {
  id: string;
  type: string;
  position?: { x: number; y: number };
  config?: Record<string, unknown>;
  config_override_schema?: Record<string, unknown>;
  label?: string;
}

File: backend/app/schemas/task.py
"""Task-related Pydantic schemas."""
from typing import Optional, List

from pydantic import BaseModel, Field


class TaskScheduleConfig(BaseModel):
    type: str = Field(default="immediate", pattern="^(immediate|cron|interval)$")
    cron_expression: Optional[str] = None
    interval_seconds: Optional[int] = Field(default=None, gt=0)


class ConfigOverrideItem(BaseModel):
    workflow_node_id: str
    agent_type_id: Optional[str] = None
    config_override: dict


class TaskCreate(BaseModel):
    name: str
    description: Optional[str] = None
    spec: Optional[str] = None
    priority: str = "medium"
    depends_on: Optional[list] = None
    assigned_agent: Optional[str] = None
    workflow_id: Optional[str] = None
    project_id: Optional[str] = None  # for standalone tasks via POST /api/v1/tasks
    config_overrides: Optional[List[ConfigOverrideItem]] = None
    schedule: Optional[TaskScheduleConfig] = None


class TaskUpdate(BaseModel):
    name: Optional[str] = None
    description: Optional[str] = None
    spec: Optional[str] = None
    priority: Optional[str] = None
    status: Optional[str] = None
    depends_on: Optional[list] = None
    assigned_agent: Optional[str] = None
    workflow_id: Optional[str] = None


class TaskOut(BaseModel):
    id: str
    project_id: str
    user_id: str
    parent_task_id: Optional[str] = None
    name: str
    description: Optional[str] = None
    spec: Optional[str] = None
    priority: str = "medium"
    status: str = "pending"
    depends_on: Optional[list] = None
    assigned_agent: Optional[str] = None
    workflow_id: Optional[str] = None
    workflow_snapshot: Optional[dict] = None
    schedule_type: Optional[str] = None
    schedule_config: Optional[dict] = None
    total_jobs: int = 0
    completed_jobs: int = 0
    total_tokens: int = 0
    total_cost: float = 0.0
    started_at: Optional[str] = None
    completed_at: Optional[str] = None
    created_at: str = ""
    updated_at: str = ""
    # enriched fields (not ORM, populated by query)
    workflow_name: Optional[str] = None
    agent_name: Optional[str] = None

    model_config = {"from_attributes": True}


class TaskAgentConfigCreate(BaseModel):
    workflow_node_id: str
    agent_type_id: Optional[str] = None
    config_override: dict


class TaskAgentConfigOut(BaseModel):
    id: str
    task_id: str
    workflow_node_id: str
    agent_type_id: Optional[str] = None
    config_override: Optional[dict] = None
    created_at: str = ""
    updated_at: str = ""

    model_config = {"from_attributes": True}
